lib/client.js

```diff
@@ -78,6 +81,11 @@ module.exports = function (dependencies) {
               this.config.token.regenerate(tokenGeneration);
               resolve(this.write(notification, device, retryCount + 1));
               return;
+            } else if (status === "500" && response.reason === "InternalServerError") {
+              stream.connection.close();
+              let error = new VError("Error 500, stream ended unexpectedly");
```

Comment:
No real concerns :) Working towards this common return pattern.

The presence of `error` helps users understand very immediately that something's gone wrong, so I see that as preferable to returning `response` (where they have to pick through status codes and messages to figure out if everything's okay or not).

If we can't specifically see an error, we're (see [lib/client.js:91](https://github.com/cmdkoh/node-apn/blob/710f68c9fb8ddd9cd71b089221ff61f7cb2a8fd0/lib/client.js#L91)) sending along the `response` just in case.

`status` is _always_ useful to have, though (like you say), even on success.

My proposal in changes:
[lib/client.js:76](https://github.com/cmdkoh/node-apn/blob/710f68c9fb8ddd9cd71b089221ff61f7cb2a8fd0/lib/client.js#L76) -> `resolve({ device, status })`
[lib/client.js:87](https://github.com/cmdkoh/node-apn/blob/710f68c9fb8ddd9cd71b089221ff61f7cb2a8fd0/lib/client.js#L87) -> `resolve({ device, status, error })`
[lib/client.js:91](https://github.com/cmdkoh/node-apn/blob/710f68c9fb8ddd9cd71b089221ff61f7cb2a8fd0/lib/client.js#L91) -> `resolve({ device, status, response })`
[lib/client.js:94](https://github.com/cmdkoh/node-apn/blob/710f68c9fb8ddd9cd71b089221ff61f7cb2a8fd0/lib/client.js#L94) -> `resolve({ device, status, error })`

The lack of `status` on any other resolutions ([lib/client.js:109](https://github.com/cmdkoh/node-apn/blob/710f68c9fb8ddd9cd71b089221ff61f7cb2a8fd0/lib/client.js#L109) and [lib/client.js:115](https://github.com/cmdkoh/node-apn/blob/710f68c9fb8ddd9cd71b089221ff61f7cb2a8fd0/lib/client.js#L115)) could then go to indicate a more internal error and a distinct lack of communication with the APNS.